File: 145.二叉树的后序遍历.py
#
# @lc app=leetcode.cn id=145 lang=python3
#
# [145] 二叉树的后序遍历
#
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        node : TreeNode = root
        stack : List = []
        res : List = []
        while node or len(stack)!=0:
            while node:
                stack.append(node)
                node = node.left
            if len(stack)!=0:
                node = stack.pop()
                if node.right:
                    stack.append(node)
                    node = node.right
                    pass
                else:
                    res.append(node.val)
                    print(node.val)
                    
                    

File: 145.二叉树的后序遍历.js
/*
 * @lc app=leetcode.cn id=145 lang=javascript
 *
 * [145] 二叉树的后序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal = function (root) {
    let node = root;
    let res = [];
    let stack = [];
    let prev = null;
    while (node || stack.length > 0) {
        while (node) {
            stack.push(node);
            node = node.left;
        }
        node = stack.pop();
        if (node.right) {
            if (node.right === prev) {
                res.push(node.val)
                prev = node;
                node = null;
            } else {
                stack.push(node)
                node = node.right
            }
        } else {
            res.push(node.val);
            prev = node;
            node = null;
        }
    }
    return res;
};
// @lc code=end
